File: src/drive/permanent-errors.ts
// Classifies Google Drive API errors into "permanent" (retrying won't help)
// and "transient" (retrying on --resume may succeed). Permanent failures are
// stored in nodes.last_error with a PERMANENT_ERROR_PREFIX marker so that
// buildWorkList() skips them on subsequent --resume runs.

export type PermanentErrorReason =
  | 'appNotAuthorizedToFile'
  | 'forbidden'
  | 'fileNotDownloadable';

export interface PermanentErrorInfo {
  permanent: true;
  reason: PermanentErrorReason;
  httpStatus: number;
}

export const PERMANENT_ERROR_PREFIX = '[permanent:';

const PERMANENT_REASONS: ReadonlySet<PermanentErrorReason> = new Set([
  'appNotAuthorizedToFile',
  'forbidden',
  'fileNotDownloadable',
]);

interface GoogleErrorDetail {
  reason?: string;
}
interface GoogleErrorBody {
  code?: number;
  errors?: GoogleErrorDetail[];
}
interface GaxiosLikeShape {
  response?: { data?: { error?: GoogleErrorBody } };
  code?: number | string;
  status?: number;
  message?: string;
}

function asRecord(value: unknown): Record<string, unknown> | null {
  return typeof value === 'object' && value !== null
    ? (value as Record<string, unknown>)
    : null;
}

function readStructured(err: unknown): {
  reason: string | null;
  status: number | null;
} {
  const root = asRecord(err);
  if (!root) return { reason: null, status: null };
  const shape = root as unknown as GaxiosLikeShape;
  const body = shape.response?.data?.error;
  const detail = body?.errors?.[0];
  const reason = typeof detail?.reason === 'string' ? detail.reason : null;
  const status =
    typeof body?.code === 'number'
      ? body.code
      : typeof shape.status === 'number'
        ? shape.status
        : typeof shape.code === 'number'
          ? shape.code
          : null;
  return { reason, status };
}

function readMessage(err: unknown): string {
  if (err instanceof Error) return err.message;
  if (typeof err === 'string') return err;
  return String(err);
}

// Google's client occasionally flattens its error body into the message
// string (JSON-stringified). Fall back to a regex so we still catch the
// permanent reason when the structured shape is missing.
function matchMessageReason(message: string): PermanentErrorReason | null {
  if (message.includes('appNotAuthorizedToFile'))
    return 'appNotAuthorizedToFile';
  if (message.includes('fileNotDownloadable')) return 'fileNotDownloadable';
  return null;
}

export function classifyDriveError(err: unknown): PermanentErrorInfo | null {
  const { reason, status } = readStructured(err);
  if (reason && PERMANENT_REASONS.has(reason as PermanentErrorReason)) {
    return {
      permanent: true,
      reason: reason as PermanentErrorReason,
      httpStatus: status ?? 403,
    };
  }

  const message = readMessage(err);
  const matched = matchMessageReason(message);
  if (matched) {
    return { permanent: true, reason: matched, httpStatus: status ?? 403 };
  }
  return null;
}

export function isPermanentErrorMessage(msg: string | null): boolean {
  return msg?.startsWith(PERMANENT_ERROR_PREFIX) ?? false;
}

export function formatPermanentError(
  info: PermanentErrorInfo,
  message: string
): string {
  const firstLine = message.split('\n', 1)[0] ?? '';
  const trimmed =
    firstLine.length > 200 ? `${firstLine.slice(0, 197)}...` : firstLine;
  return `${PERMANENT_ERROR_PREFIX}${info.reason}] ${trimmed}`;
}

const HUMAN_REASONS: Record<PermanentErrorReason, string> = {
  appNotAuthorizedToFile:
    'app not authorized to read this file (re-share with your account or see README troubleshooting)',
  forbidden: 'access forbidden by Drive',
  fileNotDownloadable: 'file is not downloadable via the API',
};

export function humanizePermanentReason(reason: PermanentErrorReason): string {
  return HUMAN_REASONS[reason];
}

File: src/drive/permanent_errors.rs
//! Classifies Drive API errors into permanent vs transient. Mirrors
//! `src/drive/permanent-errors.ts` 1:1: same reasons, same prefix marker,
//! same message format. The pipeline persists permanent errors into
//! `nodes.last_error` with the prefix so resume runs skip them.

use crate::error::CliError;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PermanentReason {
    AppNotAuthorizedToFile,
    Forbidden,
    FileNotDownloadable,
}

impl PermanentReason {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::AppNotAuthorizedToFile => "appNotAuthorizedToFile",
            Self::Forbidden => "forbidden",
            Self::FileNotDownloadable => "fileNotDownloadable",
        }
    }

    fn from_reason(reason: &str) -> Option<Self> {
        match reason {
            "appNotAuthorizedToFile" => Some(Self::AppNotAuthorizedToFile),
            "forbidden" => Some(Self::Forbidden),
            "fileNotDownloadable" => Some(Self::FileNotDownloadable),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub struct PermanentInfo {
    pub reason: PermanentReason,
    pub http_status: u16,
}

pub const PERMANENT_ERROR_PREFIX: &str = "[permanent:";

/// Inspect an arbitrary error. Returns `Some` iff the error encodes one of
/// the known permanent Drive reasons either via structured Google error body
/// or via a flattened message-string fallback.
pub fn classify(err: &CliError) -> Option<PermanentInfo> {
    let msg = &err.message;
    // Best-effort: parse JSON-shaped Google errors that may have been
    // captured into the message text. The TS code does `response?.data?.error`
    // → `errors[0].reason`; in our http_error helper we put the first body
    // line into the message, which generally contains JSON if the API
    // returned a structured error.
    if let Some(reason) = extract_reason_from_json_body(msg) {
        if let Some(reason) = PermanentReason::from_reason(&reason) {
            return Some(PermanentInfo {
                reason,
                http_status: extract_status(msg).unwrap_or(403),
            });
        }
    }
    // Substring fallback for the cases where Google's client flattens the
    // error into a free-form string. Mirrors `matchMessageReason`.
    if msg.contains("appNotAuthorizedToFile") {
        return Some(PermanentInfo {
            reason: PermanentReason::AppNotAuthorizedToFile,
            http_status: extract_status(msg).unwrap_or(403),
        });
    }
    if msg.contains("fileNotDownloadable") {
        return Some(PermanentInfo {
            reason: PermanentReason::FileNotDownloadable,
            http_status: extract_status(msg).unwrap_or(403),
        });
    }
    None
}

fn extract_reason_from_json_body(msg: &str) -> Option<String> {
    // Heuristic: find the first '{' and parse from there.
    let start = msg.find('{')?;
    let body: serde_json::Value = serde_json::from_str(msg.get(start..)?).ok()?;
    let reason = body
        .get("error")?
        .get("errors")?
        .as_array()?
        .first()?
        .get("reason")?
        .as_str()?
        .to_string();
    Some(reason)
}

fn extract_status(msg: &str) -> Option<u16> {
    // Match "HTTP <status>" — what `http_error` writes.
    let idx = msg.find("HTTP ")?;
    let tail = &msg[idx + 5..];
    let end = tail
        .find(|c: char| !c.is_ascii_digit())
        .unwrap_or(tail.len());
    tail[..end].parse().ok()
}

pub fn format_marker(info: &PermanentInfo, message: &str) -> String {
    let first_line = message.lines().next().unwrap_or("");
    let trimmed: String = if first_line.chars().count() > 200 {
        let truncated: String = first_line.chars().take(197).collect();
        format!("{truncated}...")
    } else {
        first_line.to_string()
    };
    format!(
        "{PERMANENT_ERROR_PREFIX}{}] {trimmed}",
        info.reason.as_str(),
    )
}

pub fn is_permanent_marker(msg: &str) -> bool {
    msg.starts_with(PERMANENT_ERROR_PREFIX)
}

pub fn humanize(reason: PermanentReason) -> &'static str {
    match reason {
        PermanentReason::AppNotAuthorizedToFile => {
            "app not authorized to read this file (re-share with your account or see README troubleshooting)"
        }
        PermanentReason::Forbidden => "access forbidden by Drive",
        PermanentReason::FileNotDownloadable => "file is not downloadable via the API",
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::error::ErrorCode;

    fn err(message: &str) -> CliError {
        CliError::new(message, ErrorCode::Unknown)
    }

    #[test]
    fn classifies_structured_app_not_authorized() {
        let json = r#"{"error":{"code":403,"errors":[{"reason":"appNotAuthorizedToFile"}]}}"#;
        let e = err(&format!("files.get: HTTP 403: {json}"));
        let info = classify(&e).unwrap();
        assert_eq!(info.reason, PermanentReason::AppNotAuthorizedToFile);
        assert_eq!(info.http_status, 403);
    }

    #[test]
    fn classifies_unstructured_message_fallback() {
        let e = err("download failed: appNotAuthorizedToFile somewhere in this string");
        let info = classify(&e).unwrap();
        assert_eq!(info.reason, PermanentReason::AppNotAuthorizedToFile);
    }

    #[test]
    fn ignores_transient_errors() {
        let e = err("network timeout");
        assert!(classify(&e).is_none());
        let e2 = err("rate limited");
        assert!(classify(&e2).is_none());
    }

    #[test]
    fn formats_and_detects_marker() {
        let info = PermanentInfo {
            reason: PermanentReason::Forbidden,
            http_status: 403,
        };
        let marker = format_marker(&info, "access forbidden by Drive\nsecond line");
        assert_eq!(marker, "[permanent:forbidden] access forbidden by Drive");
        assert!(is_permanent_marker(&marker));
        assert!(!is_permanent_marker("plain error"));
    }

    #[test]
    fn truncates_long_messages_in_marker() {
        let info = PermanentInfo {
            reason: PermanentReason::Forbidden,
            http_status: 403,
        };
        let long = "x".repeat(300);
        let marker = format_marker(&info, &long);
        assert!(marker.starts_with("[permanent:forbidden] "));
        // 197 + 3 ellipsis + prefix
        assert!(marker.contains("..."));
    }
}
